server/eventHandlers/slashEvents.ts

```diff
@@ -0,0 +1,51 @@
+import { IEventHandler, CWEvent, IChainEventData, SubstrateTypes } from '@commonwealth/chain-events';
+
+
+export default class extends IEventHandler {
+  constructor(
+    private readonly _models
+  ) {
+    super();
+  }
+
+  /**
+    * Event handler for slash information of validators details in DB.
+  */
+  public async handle(event: CWEvent < IChainEventData > ,dbEvent) {
+    // 1) if other event type ignore and do nothing.
+    if (event.data.kind !== SubstrateTypes.EventKind.Slash) {
+      return dbEvent;
+    }
+    const newSlashEventData = event.data;
+
+    // 2) check for the latest historical validator data from db
+    const latestValidators = await this._models.HistoricalValidatorStatistic.findOne({
+      where: {
+        stash: newSlashEventData.validator
+      },
+      order: [
+        ['created_at', 'DESC']
+      ]
+    });
+    if (!latestValidators) {
+      return dbEvent;
+    }
+
+    let validator = JSON.parse(JSON.stringify(latestValidators));
+
+    // 3) Modify new exposures for validators
+    validator.exposure.own = (Number(validator.exposure.own) - Number(newSlashEventData.amount)).toString();
```

Comment:
slash events don't only slash the validators `own` stake but also an amount from each `others` exposure. The slash event should have a more complicated structure than just an amount afaik?

Reply:
Yeah, I had this nagging feeling about this one. I knew that a slash is to be applied to both the validator as well as the norminator but just didn't know exactly how?.  As I see it there are 2 ways to do it that is either I subtract the slashed amount(given in the slash event) from validator's own and others exposure as is or I first compute the proportion of slash (slash_amount/total_stake) and then subtract this percent from the validator's own amount as well as all the other exposures.

The Later approach makes more sense to me. Consider this case study that a validator is elected for a total stake of 100 Edge in which the `own` stake is 50 and others stake is 50(20,20,10 Edge). If the slash is of 10 Edge then:
- Approach 1 will leave the validator with 40 Edge and our nominators with 10,10 and 0 Edge respectively. This will bring down the total stake down to 60
- Approach 2 will leave the validator with 45 Edge and our nominators with 18, 18, 9 Edge. This will reduce the total stake to 90.

Let me know If I am right about it or something bigger is brewing behind the scenes!

Reply:
Actually sorry! Looks like the slash event will be applied multiple times but always has AccountID and Balance
https://github.com/hicommonwealth/substrate/blob/master/frame/staking/src/lib.rs#L1183

Reply:
This make it slightly more difficult because we will receive slash events for nominators, but not have a mapping for the validators they are nominating in order to update those exposures. I will think on this tonight and if you have a solution too we can review.

Reply:
I think we can backtrack nominators from the validator using the `new-session` event (https://github.com/hicommonwealth/chain-events/blob/master/src/substrate/filters/enricher.ts#L113-L139). as we are adding `new-session`'s validators record in the `HistoricalValidatorStatistic` DB so we can get the latest session event and their activeExposures which will have the nominators with their staked amount.

Reply:
Right, we'd need to identify which validator(s) was slashed and then backtrack against that.

Reply:
Awaiting for your feedback on which approach should I follow. IMO we should go for Approach 2. Would love to know your thoughts on this.

Reply:
I think there's still some miscommunication. And this sheds some light on what's going on. Every slash event is either for a validator or a nominator. So basically we need to check that:
- the account is a validator or a nominator
- if its a nominator, figure out which validator was slashed and correlate there
- if its a nominator and multiple of its validators are slashed, we can default to an average strategy

Nonetheless, since this seems to have some error of correctness in it. We should re-poll the exposures on the session event and always reset things there, as a means of correcting any incorrect state.

Check this out:
- https://edgeware.subscan.io/event?module=staking&event=slash
- https://polkadot.subscan.io/event?module=staking&event=slash